### PR TITLE
Search Use Case is done..?

### PR DESCRIPTION
**Pull request by jerryzhu509**
Implemented SearchParser and Controller that successfully deals with the user input. 
One to-do is to fix the problem that SearchInteractor tests are not deterministic (great observation by Taylor).

### CODE REVIEW

File: src/main/java/org/hydev/mcpm/client/arguments/parsers/SearchParser.java [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+    }
+
+    @Override
+    public String description() {
+        return "Search for a plugin in the database";
+    }
+
+    @Override
+    public void configure(Subparser parser) {
+        parser.addArgument("by").choices("name", "keyword", "command").dest("type")
```
Comment (1whatleytay) [inline src/main/java/org/hydev/mcpm/client/arguments/parsers/SearchParser.java]: Maybe by default we search by name?

File: src/main/java/org/hydev/mcpm/client/arguments/parsers/SearchParser.java [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+                .help("Specifies the Type of Search");
+        parser.addArgument("text").dest("text").nargs("+")
+                .help("Specifies the search text.");
+        parser.addArgument("-n", "--no-cache").action(Arguments.storeTrue()).dest("noCache")
+                .help("Specifies whether to use local cache or not.");
+    }
+
+    @Override
+    public void run(Namespace details, Consumer<String> log) {
+        controller.searchPackages("BY_" + details.getString("type").toUpperCase(),
```
Comment (1whatleytay) [inline src/main/java/org/hydev/mcpm/client/arguments/parsers/SearchParser.java]: `"BY_" + details.getString("type").toUpperCase()` bruh get this processing out of here. This is what the controller is for!!

**Review by 1whatleytay — APPROVED** [review]

File: src/main/java/org/hydev/mcpm/client/arguments/parsers/SearchParser.java (hunk 2 above)
Comment (1whatleytay) [reply src/main/java/org/hydev/mcpm/client/arguments/parsers/SearchParser.java:43]: This is a merge blocker IMO.

File: src/test/java/org/hydev/mcpm/client/database/PluginMockFactory.java [hunk 3]
```diff
@@ -64,4 +65,13 @@ public static DatabaseInteractor interactor(List<PluginModel> plugins) {
...
         return new DatabaseInteractor(fetcher, listener);
     }
+
+    public static List<PluginModel> generatePlugins() {
+        List<PluginModel> plugins = new ArrayList<>();
+        String[] names = {"WorldGuard"};
+        String [] descriptions = {"Protect your server!\n" +
+                "WorldGuard lets you and players guard areas of land against griefers and undesirables\n" +
+                "as well as tweak and disable various gameplay features of Minecraft."};
+        return plugins;
```
Comment (MstrPikachu) [inline src/test/java/org/hydev/mcpm/client/database/PluginMockFactory.java:75]: this returns an empty list

Comment (1whatleytay) [reply src/test/java/org/hydev/mcpm/client/database/PluginMockFactory.java:75]: The truth could not be truer. I think he said he would just push what he has and work on tests later.

**Review by MstrPikachu — APPROVED** [review]
Good just remember to fix the plugin factory stuff when you make your tests

**Pull request merged**